### PR TITLE
Dataloader

### PR DESCRIPTION
**Pull request by michaelkiper**
## Overview
This PR builds out the dataloader capability for providing both simulated or static data, or both. 

The main new components are:
1. `OrchestratorDatasetArgs`
      - This struct (class) defines the set of arguments that will be passed into the `OrchestratorDataset`.
      - The main function is that it defines if static data only, simulated data only, or both will be provided
2. `OrchestratorDataset`
      - This is an inherited class of the Pytorch `IterableDataset` so that it could function on a per-batch basis instead of on an individual datapoint level.
      - The policy is that either a simulated or static dataset will be provided at some incremental level.
      - The iterator does NOT stop as well. **It will be the caller's responsibility to terminate the iterator**
      - There are a number of stateful features for the static data batches:
            - `static_epoch`, `static_epoch_step`, `static_samples_seen`, `is_simulated_batch`
            - From these, one could then infer the simulated data info
3. `dataloader_from_config`
      - This function creates a Pytorch DataLoader with the `OrchestratorDataset` as the dataset generator, building the arguments from the config files and a set of 

1 small refactor:
- src/cover_class/utils.py `read_config`
     - This is added for the `dataloader_from_config` not reloading the config when it calls `args_from_config`

1 new stub:
- src/cover_class/simulation/simulate.py `args_from_config`
     - This is included as the `dataloader_from_config` needs to call it and receive the intended args. 

#### User's Notes
Example intended use case:
```
>>> old = OrchestratorDataset(args)
>>> dl = DataLoader(old, batch_size=None)
>>> # NOTE: The `batch_size` in the dataloader must be set to `None`
>>> for X, Y in dl:
>>>     ...
```
or
```
>>> my_config_path = '/my/path/config.yaml'
>>> dataloader: DataLoader = dataloader_from_config( # the DataLoader has the OrchestratorDataset generator
    my_config_path, 
    data_train, 
    labels_train,
    batch_size,
    shuffle
)
```

#### Misc. Notes
You will likely notice that the batch size is NOT a property of the config and needs to be passed through as a parameter. This is intentional. The reason being is that the batch size is a function of the model (training) hyperparameters and training process which requires its own study on the critical batch sizes, loss and gradient effects, etc. As such, it's much more appropriate to not include it here and allow the user to control that from their end so that way they could do a proper batch size study for their model.

## Testing
The testing conducted has been in the form of unit tests that I tried to make representative of the actual use cases that could arise. You could run these as well with `make test`

### CODE REVIEW

File: src/cover_class/dataloader/dataloader.py [hunk 1]
```diff
@@ -1,60 +1,134 @@
...
+        self.static_epoch += 1
+        self.static_epoch_step = 0
+        self.__shuffle__()
+
+    def __use_static_predicate__(self) -> bool:
+        return bool(self.args._method_selection_idxs[(self.step % 100)])
+    
+    def static_epoch_len(self) -> Optional[int]:
+        if self.args.static_data is None: return None
+        return int(len(self.args.static_data) / self.args.percent_static)
```
[the remark this answers is not in the document]
Comment (michaelkiper) [reply src/cover_class/dataloader/dataloader.py]: This isn't true. This function is meant to tell you how many batches are seen before a full epoch of the static data is seen (ik the math is off for that tho). I do think that it's redundant and unnecessary  though, so I'll remove it.

File: src/cover_class/dataloader/dataloader.py [hunk 2]
```diff
@@ -1,60 +1,134 @@
...
+            self.step += 1
+            if self.args._using_static and self.__use_static_predicate__():
+                self.is_simulated_batch = False
+                start =  (self.static_epoch_step * self.args.batch_size)
+                end   = ((self.static_epoch_step+1) * self.args.batch_size)
+                self.static_epoch_step += 1
+                # mypy doesn't catch self.args._using_static
+                idx = self._static_idx_order[start: end] # type: ignore
+                self.static_samples_seen += len(idx)
+                if end >= len(self.args.static_data)-1: # type: ignore
```
[the remark this answers is not in the document]
Comment (michaelkiper) [reply src/cover_class/dataloader/dataloader.py:83]: Ya true

**Comment by michaelkiper** [comment]
@pgbrodrick tagging you on this PR as a reviewer as well

File: src/cover_class/utils.py [hunk 3]
```diff
@@ -1,5 +1,6 @@
 from typing import Dict
-import yaml
+import yaml # type: ignore[import]
```
Comment (pgbrodrick) [inline src/cover_class/utils.py:2]: Can remove comment.

**Comment by pgbrodrick** [comment]
Reviewed, and looks reasonable - I'm not executing until we're further along.  @jakehlee , if this looks good to you, let's get it in.  @michaelkiper , does #2 need to go first?

**Review by jakehlee — APPROVED** [review]
Quick question, would there ever be a case for simulated data reproducibility? We can always set the torch random seed at top of whatever main script is calling this, but is there any benefit to implementing a seed for simulated test sets? Or would we always generate that externally and provide it as a static dataset?

Otherwise LGTM 👍

**Comment by michaelkiper** [comment]
> Quick question, would there ever be a case for simulated data reproducibility? We can always set the torch random seed at top of whatever main script is calling this, but is there any benefit to implementing a seed for simulated test sets? Or would we always generate that externally and provide it as a static dataset?
> 
> Otherwise LGTM 👍

Good question. I was under the impression that the test set will always be from the static datasets (makes sense imo). The simulations I think are only there to see if they could allow the model to improve on the real physical data. (I could be wrong though)

I do see there being merit in seeding the simulated data beforehand, but I don't think it's terribly useful to do in the scope of this PR and would likely need to be handled in the user's own codebase. The reason being is that a user could call a random function after every dataloader iteration, which messes up the reproducibility. So we probably shouldn't offer a false sense of reproducibility here.

**Comment by jakehlee** [comment]
Makes sense, ship it!

**Review by pgbrodrick — APPROVED** [review]
